app/scripts/lib/local-store.js

```diff
@@ -11,6 +12,37 @@ export default class ExtensionStore {
     if (!this.isSupported) {
       log.error('Storage local API not available.');
     }
+    // we use this flag to avoid flooding sentry with a ton of errors:
+    // once data persistence fails once and it flips true we don't send further
+    // data persistence errors to sentry
+    this.dataPersistenceFailing = false;
+  }
+
+  setMetaData(initMetaData) {
+    this.metadata = initMetaData;
+  }
+
+  async set(state) {
+    if (!state) {
+      throw new Error('MetaMask - updated state is missing');
+    }
+    if (this.isSupported) {
+      try {
+        // we format the data for storage as an object with the "data" key for the controller state object
+        // and the "meta" key for a metadata object containing a version number that tracks how the data shape
+        // has changed using migrations to adapt to backwards incompatible changes
+        await this._set({ data: state, meta: this.metadata });
```

Comment:
We're not checking here to ensure `this.metadata` is set. If it's unset, that would _severely_ mess up the persisted wallet state. It might result in the wallet being bricked/unrecoverable. We should explicitly check that and throw if it's unset before referencing it here.

Reply:
Done [here](https://github.com/MetaMask/metamask-extension/pull/15978/commits/817d85d91b67aaf401490b7372ebb7511919b37a) + added tests